README
Qore json module

INTRODUCTION
------------
The Qore json module provides JSON functionality to Qore.  This was previously
part of the main qore library but moved to this module in qore v0.8.1 

see docs for more information


BUILDING
--------
Requires qore 0.8.1+ and openssl headers and libraries to build and run

to configure the build, execute
   	configure --disable-debug

if the qore library cannot be found, then you can use:
   	configure --disable-debug --with-qore=<dir>

if openssl cannot be found, then you can use:
        configure --disable-debug --with-openssl-dir=<dir>

The qore binary also needs to be in the path so configure can determine the module directory

Then execute

make && make install

(or 'make && sudo make install' as needed)

please direct any questions to:
david@qore.org

or you can post in the qore forums at
        http://qore.org/forums


ChangeLog
2013-09-30 15:37  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl, src/ql_json.qpp:
	  date/time values are always serialized with microseconds and in
	  the local time zone for consistency's sake

2013-09-24 15:30  david_nichols

	* qlib/JsonRpcHandler.qm: fixed typo in last commit

2013-09-24 14:56  david_nichols

	* qlib/JsonRpcHandler.qm: cosmetic/doc updates

2013-09-18 15:30  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl,
	  qlib/JsonRpcHandler.qm: updated the XmlRpcHandler module for
	  enhanced logging

2013-09-08 19:04  david_nichols

	* src/QC_JsonRpcClient.qpp: fixed to compile with new
	  QoreHttpClientObject class if available

2013-09-05 14:26  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-json-module.spec, src/ql_json.qpp: updated module to v1.4 -
	  fixed JSON control character escaping so that only the string
	  being concatenated is processed and also that only ASCII control
	  characters are processed (and not bytes inside a multi-byte
	  character that may look like ASCII control characters)

2013-09-05 13:54  david_nichols

	* src/ql_json.qpp: fixed json encoding of control chars

2013-08-02 06:06  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-json-module.spec, src/QC_JsonRpcClient.qpp: include
	  charset=utf-8 in Content-Type header JsonRpcClient; JSON messages
	  are always serialized with UTF-8 encoding
	  updated to v1.3

2013-07-31 11:06  david_nichols

	* ChangeLog: updated changelog for 1.2 release

2013-06-14 04:51  david_nichols

	* src/ql_json.qpp: removed commented-out code

2013-06-13 12:08  david_nichols

	* src/ql_json.qpp: non-functional changes for efficiency

2013-06-13 11:52  david_nichols

	* ChangeLog, RELEASE-NOTES, configure.ac,
	  docs/mainpage.doxygen.tmpl, qore-json-module.spec,
	  src/ql_json.qpp: corrected json encoding/decoding of
	  escape/control characters according to
	  http://tools.ietf.org/html/rfc4627

2013-02-27 10:23  david_nichols

	* RELEASE-NOTES, docs/mainpage.doxygen.tmpl: updated release notes
	  for 1.1 release

2013-02-27 10:10  david_nichols

	* README-BINARIES: updated binary location (and changelog for
	  pgsql)

2013-02-08 15:40  subik

	* src/ql_json.qpp, test/json-test.q: json: escape new line and tabs
	  (I'm not sure in its correctness but it seems it's working...)

2012-10-30 22:01  david_nichols

	* RELEASE-NOTES, src/ql_json.qpp: added support for serializing the
	  "number" type

2012-09-28 16:19  david_nichols

	* configure.ac: applied reini urban's multi-arch patches for debian
	  to all modules

2012-08-28 07:24  david_nichols

	* Makefile.am: removed old files from Makefile

2012-06-25 08:51  david_nichols

	* configure.ac: updated configure.ac logic to deal with /lib ->
	  /usr/lib and /lib64 -> /usr/lib64 as with Fedora 17 - the
	  previous logic was too clever for its own good

2012-06-08 12:59  david_nichols

	* Makefile.am, configure.ac: minor updates

2012-06-01 08:47  david_nichols

	* examples/HttpServer.qc, examples/JsonRpcServerValidation.q:
	  updated JsonRpcServerValidation.q to use user modules; removed
	  HttpServer class (supplied as a user module with qore 0.8.4+)

2012-06-01 08:18  david_nichols

	* Makefile.am, configure.ac, docs/doxyfile.tmpl,
	  docs/mainpage.doxygen.tmpl, docs/qlib, docs/qlib/Doxyfile.tmpl,
	  examples/JsonRpcHandler.qc, qlib, qlib/JsonRpcHandler.qm,
	  qore-json-module.spec, src/Makefile.am, src/QC_JsonRpcClient.cpp,
	  src/QC_JsonRpcClient.h, src/QC_JsonRpcClient.qpp,
	  src/json-module.cpp, src/ql_json.cpp, src/ql_json.h,
	  src/ql_json.qpp: updated json module to qpp

2012-01-16 13:07  david_nichols

	* configure.ac: updated configure scripts to only set LIBSUFFIX=64
	  on 64-bit platfiorms if /usr/lib64 exists

2011-09-12 16:26  david_nichols

	* configure.ac: added -mt to compile and link lines when building
	  on solaris with CC

2011-09-05 14:49  subik

	* ChangeLog: missing file for make dist

2011-07-29 11:00  david_nichols

	* configure.ac, src/Makefile.am: patches to the json module to
	  allow for building under mingw32 cross-compile
	  environment; updated xml module to delete temporary files in
	  module directory
	  when building for windows under mingw32

2010-12-28 08:31  david_nichols

	* RELEASE-NOTES: updated xml module to 1.1, updated release notes,
	  updated ssh2 changelog

2010-12-28 08:29  david_nichols

	* configure.ac, qore-json-module.spec: updated to version 1.1

2010-12-28 08:26  david_nichols

	* src/QC_JsonRpcClient.cpp: doc updates

2010-12-26 20:44  david_nichols

	* docs/doxygen-api-filter.q, examples/json-rpc-client.q: doc and
	  example updates

2010-12-23 21:03  david_nichols

	* Makefile.am, README, configure.ac, qore-json-module.spec:
	  packaging fixes - use bz2, do not include .svn dirs

2010-12-23 20:20  david_nichols

	* qore-json-module.spec: spec file fixes

2010-12-23 07:59  david_nichols

	* Makefile.am, qore-json-module.spec: fixed rpm building -
	  including doc rpms for the xml and json modules

2010-12-22 20:03  david_nichols

	* src/ql_json.cpp: documentation updates

2010-12-22 18:37  david_nichols

	* docs/mainpage.doxygen.tmpl: minor doc updates

2010-12-22 18:17  david_nichols

	* ., AUTHORS, COPYING, Makefile.am, README, README-BINARIES,
	  RELEASE-NOTES, configure.ac, docs, docs/doxyfile.tmpl,
	  docs/doxygen-api-filter.q, docs/footer_template.html,
	  docs/mainpage.doxygen.tmpl, examples, examples/HttpServer.qc,
	  examples/JsonRpcHandler.qc, examples/JsonRpcServerValidation.q,
	  examples/json-rpc-client.q, m4, m4/acx_pthread.m4,
	  qore-json-module.spec, reconf.sh, src, src/Makefile.am,
	  src/QC_JsonRpcClient.cpp, src/QC_JsonRpcClient.h,
	  src/json-module.cpp, src/ql_json.cpp, src/ql_json.h,
	  src/qore-json-module.h, src/single-compilation-unit.cpp, test,
	  test/json-test.q: xml module updates, separated json
	  functionality out to json module



src/QC_JsonRpcClient.cpp
/*
  QC_JsonRpcClient.cpp

  Qore Programming Language

  Copyright (C) 2006 - 2010 Qore Technologies

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

//! @file JsonRpcClient.qc defines the JsonRpcClient class

#include "qore-json-module.h"

#include <qore/QoreHTTPClient.h>
#include <qore/ReferenceHolder.h>

#include "ql_json.h"

qore_classid_t CID_JSONRPCCLIENT;
QoreClass *QC_JSONRPCCLIENT;

class HTTPInfoRefHelper {
protected:
   const ReferenceNode *ref;
   ExceptionSink *xsink;
   ReferenceHolder<QoreHashNode> info;

public:
   DLLLOCAL HTTPInfoRefHelper(const ReferenceNode *n_ref, QoreStringNode *msg, ExceptionSink *n_xsink) : ref(n_ref), xsink(n_xsink), info(new QoreHashNode, xsink) {
      info->setKeyValue("request", msg, xsink);
   }
   DLLLOCAL ~HTTPInfoRefHelper() {
      // we have to create a temporary ExceptionSink if there is
      // an active exception, otherwise writing back the reference will fail
      ExceptionSink *txsink = *xsink ? new ExceptionSink : xsink;
      
      // write info hash to reference
      AutoVLock vl(txsink);
      QoreTypeSafeReferenceHelper rh(ref, vl, txsink);
      if (!rh)
         return;

      if (rh.assign(info.release(), txsink))
         return;   

      if (txsink != xsink)
         xsink->assimilate(txsink);
   }
   DLLLOCAL QoreHashNode *operator*() {
      return *info;
   }
};

static void set_jrc_defaults(QoreHTTPClient &client) {
   // set encoding to UTF-8
   client.setEncoding(QCS_UTF8);

   // set options for JSON-RPC communication
   client.setDefaultPath("JSON");
   client.setDefaultHeaderValue("Content-Type", "application/json");
   client.setDefaultHeaderValue("Accept", "application/json");
   client.setDefaultHeaderValue("User-Agent", "Qore JSON-RPC Client v" PACKAGE_VERSION);

   client.addProtocol("jsonrpc", 80, false);
   client.addProtocol("jsonrpcs", 443, true);
}

//! main %Qore Programming Language namespace
/**# namespace Qore {
 */
//! The JsonRpcClient class provides easy access to JSON-RPC web services
/** This class inherits all public methods of the HTTPClient class. The inherited HTTPClient methods are not listed in this section, see the documentation for the HTTPClient class for more information on methods provided by the parent class.  For a list of low-level JSON-RPC functions, see @ref JSONRPC.

    The JsonRpcClient class understands the following protocols in addition to the protocols supported by the HTTPClient class:

    <b>JsonRpcClient Class Protocols</b>
    |!Protocol|!Default Port|!SSL?|!Description
    |\c jsonrpc|\c 80|No|Unencrypted JSON-RPC protocol over HTTP
    |\c jsonrpcs|\c 443|Yes|JSON-RPC protocol over HTTP with SSL/TLS encryption

    The JsonRpcClient supplies default values for HTTP headers as follows:

    <b>JsonRpcClient Default, but Overridable Headers</b>
    |!Header|!Default Value
    |\c Accept|\c text/json
    |\c Content-Type|\c text/json
    |\c User-Agent|\c Qore JSON-RPC Client v1.0
    |\c Connection|\c Keep-Alive

    @note This class is not available with the PO_NO_NETWORK parse option.
*/
/**# class JsonRpcClient : Qore::HTTPClient {
public:
*/

//! Creates the JsonRpcClient object based on the parameters passed
/** No connection is made because no connection parameters are set with this call; connection parameters must be set afterwards using the appropriate HTTPClient methods.
    @par Example:
    @code
my JsonRpcClient $jrc();
$jrc.setURL("http://localhost:8080");@endcode
*/
//# constructor() {}
static void JRC_constructor(QoreObject *self, const QoreListNode *params, ExceptionSink *xsink) {
   // get HTTPClient object
   ReferenceHolder<QoreHTTPClient> client((QoreHTTPClient *)self->getReferencedPrivateData(CID_HTTPCLIENT, xsink), xsink);
   if (!client)
      return;

   set_jrc_defaults(*(*client));
}

//! Creates the JsonRpcClient object based on the parameters passed
/** By default the object will immediately attempt to establish a connection to the server
    @param $opts HTTPClient constructor options:
    - \c url: A string giving the URL to connect to.
    - \c default_port: The default port number to connect to if none is given in the URL.
    - \c protocols: A hash describing new protocols, the key is the protocol name and the value is either an integer giving the default port number or a hash with \c 'port' and \c 'ssl' keys giving the default port number and a boolean value to indicate that an SSL connection should be established.
    - \c http_version: Either \c '1.0' or \c '1.1' for the claimed HTTP protocol version compliancy in outgoing message headers.
    - \c default_path: The default path to use for new connections if a path is not otherwise specified in the connection URL.
    - \c max_redirects: The maximum number of redirects before throwing an exception (the default is 5).
    - \c proxy: The proxy URL for connecting through a proxy.
    - \c timeout: The timeout value in milliseconds (also can be a relative date-time value for clarity, ex: \c 5m)
    - \c connect_timeout: The timeout value in milliseconds for establishing a new socket connection (also can be a relative date-time value for clarity, ex: \c 30s)
    @param $no_connect pass a boolean True value argument to suppress the automatic connection and establish a connection on demand with the first request
    @see HTTPClient::constructor() and HTTPClient::connect() for information on possible exceptions
    @par Example:
    @code my JsonRpcClient $jrc(("url": "http://authuser:authpass@otherhost:8080/JSONRPC"));@endcode
*/
//# constructor(hash $opts, softbool $no_connect = False) {}
static void JRC_constructor_hash_bool(QoreObject *self, const QoreListNode *params, ExceptionSink *xsink) {
   // get HTTPClient object
   ReferenceHolder<QoreHTTPClient> client((QoreHTTPClient *)self->getReferencedPrivateData(CID_HTTPCLIENT, xsink), xsink);
   if (!client)
      return;

   set_jrc_defaults(*(*client));

   const QoreHashNode* n = HARD_QORE_HASH(params, 0);
   if (client->setOptions(n, xsink))
      return;

   // do not connect immediately if the second argument is True
   if (!HARD_QORE_BOOL(params, 1))
      client->connect(xsink); 
}

//! Throws an exception; copying JsonRpcClient objects is currently not supported
/** @throw JSONRPCCLIENT-COPY-ERROR copying JsonRpcClient objects is currently not supported
*/
//# Qore::JsonRpcClient copy() {}
static void JRC_copy(QoreObject *self, QoreObject *old, QoreHTTPClient* client, ExceptionSink *xsink) {
   xsink->raiseException("JSONRPCCLIENT-COPY-ERROR", "copying JsonRpcClient objects is not yet supported.");
}

static AbstractQoreNode *make_jsonrpc_call(QoreHTTPClient *client, QoreStringNode *msg, QoreHashNode *info, ExceptionSink *xsink) {
   ReferenceHolder<QoreHashNode> response(client->send("POST", 0, 0, msg->getBuffer(), msg->strlen(), true, info, xsink), xsink);
   if (!response)
      return 0;

   ReferenceHolder<AbstractQoreNode> ans(response->takeKeyValue("body"), xsink);

   if (!ans)
      return 0;

   AbstractQoreNode *ah = *ans;
   if (info) {
      info->setKeyValue("response", ans.release(), xsink);
      info->setKeyValue("response_headers", response.release(), xsink);
   }
   
   if (ah->getType() != NT_STRING) {
      xsink->raiseException("JSONRPCCLIENT-RESPONSE-ERROR", "undecoded binary response received from remote server");
      return 0;
   }

   // parse JSON-RPC response   
   return parseJSONValue(reinterpret_cast<QoreStringNode *>(ah), xsink);
}

static AbstractQoreNode *JRC_callArgs(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   // create the outgoing message in JSON-RPC call format
   SimpleRefHolder<QoreStringNode> msg(makeJSONRPC11RequestStringArgs(params, xsink));
   if (!msg)
      return 0;

   return make_jsonrpc_call(client, *msg, 0, xsink);
}

//! Calls a remote method using a single value after the method name for the method arguments and returns the response as qore data structure
/** @param $method The JSON-RPC method name to call
    @param $args An optional list of arguments (or single argument) for the method

    @return a data structure corresponding to the JSON data returned by the JSON-RPC server

    @throw JSON-SERIALIZATION-ERROR cannot serialize value passed (ex: binary, object)
    @throw JSON-PARSE-ERROR syntax error parsing JSON string
    @throw HTTP-CLIENT-TIMEOUT timeout on response from HTTP server
    @throw HTTP-CLIENT-RECEIVE-ERROR error communicating with HTTP server

    @note other exceptions may be thrown related to communication errors (ex: SSL errors, etc)

    @par Example:
    @code my hash $result = $jrc.callArgs("method.name", $arg_list); @endcode
*/
//# hash callArgs(string $method, any $args) {}
static AbstractQoreNode *JRC_call(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   // create the outgoing message in JSON-RPC call format
   SimpleRefHolder<QoreStringNode> msg(makeJSONRPC11RequestString(params, xsink));
   if (!msg)
      return 0;

   return make_jsonrpc_call(client, *msg, 0, xsink);
}

//! Calls a remote method using a single value after the method name for the method arguments and returns the response as qore data structure, accepts a reference to a hash as the first argument to give technical information about the call
/** @param $info a reference to a hash that provides the following keys on output giving technical information about the HTTP call:
    - \c request: the literal outgoing request body sent
    - \c request-uri: the first line of the HTTP request
    - \c headers: a hash of HTTP headers in the outgoing request
    - \c response-uri: the first line of the HTTP response
    - \c response: the literal response body received from the server
    - \c response_headers: a hash of headers received in the response
    @param $method The JSON-RPC method name to call
    @param $args An optional list of arguments (or single argument) for the method

    @return a data structure corresponding to the JSON data returned by the JSON-RPC server

    @throw JSON-SERIALIZATION-ERROR cannot serialize value passed (ex: binary, object)
    @throw JSON-PARSE-ERROR syntax error parsing JSON string
    @throw HTTP-CLIENT-TIMEOUT timeout on response from HTTP server
    @throw HTTP-CLIENT-RECEIVE-ERROR error communicating with HTTP server

    @note other exceptions may be thrown related to communication errors (ex: SSL errors, etc)

    @par Example:
    @code
my hash $info;
my hash $result = $jrc.callArgsWithInfo(\$info, "method.name", $arg_list);@endcode
*/
//# hash callArgsWithInfo(reference $info, string $method, any $args) {}
static AbstractQoreNode *JRC_callArgsWithInfo(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   // get info reference
   const ReferenceNode *ref = HARD_QORE_REF(params, 0);

   // get arguments
   ReferenceHolder<QoreListNode> args(params->copyListFrom(1), xsink);

   // create the outgoing message in JSON-RPC call format
   QoreStringNode *msg = makeJSONRPC11RequestStringArgs(*args, xsink);
   if (!msg)
      return 0;

   HTTPInfoRefHelper irh(ref, msg, xsink);

   // send the message to the server and get the response as a JSON string
   ReferenceHolder<AbstractQoreNode> rv(make_jsonrpc_call(client, msg, *irh, xsink), xsink);

   return *xsink ? 0 : rv.release();
}

//! Calls a remote method taking all arguments after the method name for the method arguments and returns the response as qore data structure, accepts a reference to a hash as the first argument to give technical information about the call
/** @param $info a reference to a hash that provides the following keys on output giving technical information about the HTTP call:
    - \c request: the literal outgoing request body sent
    - \c request-uri: the first line of the HTTP request
    - \c headers: a hash of HTTP headers in the outgoing request
    - \c response-uri: the first line of the HTTP response
    - \c response: the literal response body received from the server
    - \c response_headers: a hash of headers received in the response
    @param $method The JSON-RPC method name to call

    @return a data structure corresponding to the JSON data returned by the JSON-RPC server

    @throw JSON-SERIALIZATION-ERROR cannot serialize value passed (ex: binary, object)
    @throw JSON-PARSE-ERROR syntax error parsing JSON string
    @throw HTTP-CLIENT-TIMEOUT timeout on response from HTTP server
    @throw HTTP-CLIENT-RECEIVE-ERROR error communicating with HTTP server

    @note other exceptions may be thrown related to communication errors (ex: SSL errors, etc)

    @par Example:
    @code
my hash $info;
my hash $result = $jrc.callWithInfo(\$info, "method.name", $arg1, $arg2);@endcode
*/
//# hash callWithInfo(reference $info, string $method, ...) {}
static AbstractQoreNode *JRC_callWithInfo(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   // get info reference
   const ReferenceNode *ref = HARD_QORE_REF(params, 0);

   // get arguments
   ReferenceHolder<QoreListNode> args(params->copyListFrom(1), xsink);

   // create the outgoing message in JSON-RPC call format
   QoreStringNode *msg = makeJSONRPC11RequestString(*args, xsink);
   if (!msg)
      return 0;

   HTTPInfoRefHelper irh(ref, msg, xsink);

   // send the message to the server and get the response as a JSON string
   ReferenceHolder<AbstractQoreNode> rv(make_jsonrpc_call(client, msg, *irh, xsink), xsink);

   return *xsink ? 0 : rv.release();
}

//! clears the event queue for the JsonRpcClient object
/** @par Example:
    @code $jrc.setEventQueue(); @endcode
 */
//# nothing setEventQueue() {}
static AbstractQoreNode *JRC_setEventQueue_nothing(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   client->setEventQueue(0, xsink);
   return 0;
}

//! sets the event queue for the JsonRpcClient object
/** @param $queue the Queue object to receive network events from the JsonRpcClient object
    @par Example:
    @code
my Queue $queue();
$jrc.setEventQueue($queue); @endcode
 */
//# nothing setEventQueue(Queue $queue) {}
static AbstractQoreNode *JRC_setEventQueue_queue(QoreObject *self, QoreHTTPClient *client, const QoreListNode *params, ExceptionSink *xsink) {
   HARD_QORE_OBJ_DATA(q, Queue, params, 0, CID_QUEUE, "Queue", "JsonRpcClient::setEventQueue", xsink);
   if (*xsink)
      return 0;
   // pass reference from QoreObject::getReferencedPrivateData() to function
   client->setEventQueue(q, xsink);
   return 0;
}

/**#
};
};
*/

QoreClass *initJsonRpcClientClass(QoreClass *http_client) {
   assert(QC_QUEUE);

   QC_JSONRPCCLIENT = new QoreClass("JsonRpcClient", QDOM_NETWORK);
   CID_JSONRPCCLIENT = QC_JSONRPCCLIENT->getID();

   QC_JSONRPCCLIENT->addDefaultBuiltinBaseClass(http_client);

   QC_JSONRPCCLIENT->setConstructorExtended(JRC_constructor);
   QC_JSONRPCCLIENT->setConstructorExtended(JRC_constructor_hash_bool, false, QC_NO_FLAGS, QDOM_DEFAULT, 2, hashTypeInfo, QORE_PARAM_NO_ARG, softBoolTypeInfo, &False);

   QC_JSONRPCCLIENT->setCopy((q_copy_t)JRC_copy);
   QC_JSONRPCCLIENT->addMethodExtended("callArgs",         (q_method_t)JRC_callArgs, false, QC_NO_FLAGS, QDOM_DEFAULT, 0, 2, stringTypeInfo, QORE_PARAM_NO_ARG, anyTypeInfo, QORE_PARAM_NO_ARG);
   QC_JSONRPCCLIENT->addMethodExtended("call",             (q_method_t)JRC_call, false, QC_USES_EXTRA_ARGS, QDOM_DEFAULT, 0, 1, stringTypeInfo, QORE_PARAM_NO_ARG);
   QC_JSONRPCCLIENT->addMethodExtended("callArgsWithInfo", (q_method_t)JRC_callArgsWithInfo, false, QC_NO_FLAGS, QDOM_DEFAULT, 0, 3, referenceTypeInfo, QORE_PARAM_NO_ARG, stringTypeInfo, QORE_PARAM_NO_ARG, anyTypeInfo, QORE_PARAM_NO_ARG);
   QC_JSONRPCCLIENT->addMethodExtended("callWithInfo",     (q_method_t)JRC_callWithInfo, false, QC_USES_EXTRA_ARGS, QDOM_DEFAULT, 0, 2, referenceTypeInfo, QORE_PARAM_NO_ARG, stringTypeInfo, QORE_PARAM_NO_ARG);

   // nothing JsonRpcClient::setEventQueue()  
   QC_JSONRPCCLIENT->addMethodExtended("setEventQueue",    (q_method_t)JRC_setEventQueue_nothing, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo);
   // nothing JsonRpcClient::setEventQueue(Queue $queue)  
   QC_JSONRPCCLIENT->addMethodExtended("setEventQueue",    (q_method_t)JRC_setEventQueue_queue, false, QC_NO_FLAGS, QDOM_DEFAULT, nothingTypeInfo, 1, QC_QUEUE->getTypeInfo(), QORE_PARAM_NO_ARG);

   return QC_JSONRPCCLIENT;
} 
